project-koku/koku
# [COST-2448] OCP GCP Tag Cost Duping on Multi-Cluster setup with overlapping tags

Fixes [COST-2448](https://issues.redhat.com/browse/COST-2448)
Issue Context:
* A multi cluster setup with an overlapping tag key:value pair leads to costs being multiplicatively higher depending on the number of clusters that share that key value pair

Changes proposed in this PR:
* common.py
          1. changed the ocp_matched tag check to be more specific and check for the cluster_id or cluster_alias. before it was picking up anything with kubernetes-io-cluster- and this way it will be specific to the current cluster
          2. on the special case tag matching, i am checking the string for openshift_cluster followed by cluster_id or cluster_alias so that we specifically grab ones relevant to this cluster 

* reporting_ocpgcpcostlineitem_daily_summary.sql
      1. Included in the where clause a check on the cluster id to make sure it matches the current cluster id, we add that to the gcp data during resource matching so we might as well use it: https://github.com/project-koku/koku/blob/main/koku%2Fmasu%2Futil%2Fgcp%2Fcommon.py#L194
      2. comment out the matched tag matching which we will need to look into and discuss further about the way forward to reintroduce this as multi-cluster setups with overlapping tags were leading to multiplication of costs

Testing Instructions:
1. Bring up koku with trino enabled, such as: `make docker-up-min-trino`
2. Ingest multiple OCP clusters data that share a tag key:value pair, one way to do this is to send the same ocp yml to a few different clusters with nise and then ingest the separate clusters, example yml: [ocp_cluster_1.yml.txt](https://github.com/project-koku/koku/files/8214798/ocp_cluster_1.yml.txt):
```
nise report ocp --static-report-file ocp_cluster_1.yml --ocp-cluster-id ocp_cluster_1 --insights-upload testing/pvc_dir/insights_local 
nise report ocp --static-report-file ocp_cluster_1.yml --ocp-cluster-id ocp_cluster_2--insights-upload testing/pvc_dir/insights_local 
nise report ocp --static-report-file ocp_cluster_1.yml --ocp-cluster-id ocp_cluster_3 --insights-upload testing/pvc_dir/insights_local 
```
5. Ingest a GCP source that shares the key:value pair on some compute engine generators, example yml: [gcp_static_data.yml.txt](https://github.com/project-koku/koku/files/8214796/gcp_static_data.yml.txt)
6. Wait for processing to finish and check the costs for ocp on gcp vs the costs for gcp either via endpoints or the db:
```
SELECT SUM(unblended_cost) from reporting_ocpgcpcostlineitem_daily_summary_p;

SELECT SUM(unblended_cost) from reporting_gcpcostentrylineitem_daily_summary;
```
7. if you are on main, the costs for the two will be vastly different, the ocp on gcp costs will be significantly higher. On this branch they will only be a little higher, there is an underlying issue that is being looked into separately:  [COST-####](https://issues.redhat.com/browse/COST-####)

---
Additional Context:
